src/main/scala/net/snowflake/spark/snowflake/SnowflakeRelation.scala

```diff
@@ -147,59 +149,85 @@ private[snowflake] case class SnowflakeRelation(
       }
     } else {
       // Unload data from Snowflake into a temporary directory in S3:
-      //  val tempDir = params.createPerQueryTempDir()
-      // val unloadSql =
-      //     buildUnloadStmt(standardQuery(requiredColumns, filters), tempDir)
       val prunedSchema = pruneSchema(schema, requiredColumns)
 
-      getRDDFromS3[Row](standardQuery(requiredColumns, filters), prunedSchema)
+      getRDD[Row](standardQuery(requiredColumns, filters), prunedSchema)
     }
   }
 
   // Get an RDD from an unload statement. Provide result schema because
   // when a custom SQL statement is used, this means that we cannot know the results
   // without first executing it.
-  private def getRDDFromS3[T: ClassTag](sql: String,
-                                        resultSchema: StructType): RDD[T] = {
+  private def getRDD[T: ClassTag](
+                                   sql: String,
+                                   resultSchema: StructType,
+                                   format: SupportedFormat = SupportedFormat.CSV
+                                 ): RDD[T] = {
+    val source: SupportedSource =
+      if (params.usingExternalStage) SupportedSource.EXTERNAL
+      else SupportedSource.INTERNAL
 
-    if (params.usingExternalStage) {
-      val tempDir = params.createPerQueryTempDir()
+    val rdd: RDD[String] = io.readRDD(sqlContext, params, sql, jdbcWrapper, source, format)
 
-      val numRows = setup(
-        sql = buildUnloadStmt(
-          query = sql,
-          location = Utils.fixUrlForCopyCommand(tempDir),
-          compression = if (params.sfCompress) "gzip" else "none",
-          credentialsString = Some(
-            CloudCredentialsUtils.getSnowflakeCredentialsString(sqlContext,
-                                                              params))),
+    format match {
```

Comment:
I don't like having this inline here like this. Can we move it to a new class, like CSVFormat converter?

Reply:
Sure, I can create new functions to handles each format